### PR TITLE
[SPARK-25680][SQL] SQL execution listener shouldn't happen on execution thread

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

The SQL execution listener framework was created from scratch(see https://github.com/apache/spark/pull/9078). It didn't leverage what we already have in the spark listener framework, and one major problem is, the listener runs on the spark execution thread, which means a bad listener can block spark's query processing.

This PR re-implements the SQL execution listener framework. Now `ExecutionListenerManager` is just a normal spark listener, which watches the `SparkListenerSQLExecutionEnd` events and post events to the 
user-provided SQL execution listeners.

## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gatorsmile @brkyvz

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 1]
```diff
@@ -39,7 +39,14 @@ case class SparkListenerSQLExecutionStart(
 
 @DeveloperApi
 case class SparkListenerSQLExecutionEnd(executionId: Long, time: Long)
-  extends SparkListenerEvent
+  extends SparkListenerEvent {
+
+  @JsonIgnore private[sql] var executionName: Option[String] = None
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: For backward compatibility, I make these new fields `var`.

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala [hunk 2]
```diff
@@ -75,95 +76,70 @@ trait QueryExecutionListener {
...
   }
 
   /**
    * Get an identical copy of this listener manager.
    */
   @DeveloperApi
-  override def clone(): ExecutionListenerManager = writeLock {
-    val newListenerManager = new ExecutionListenerManager
-    listeners.foreach(newListenerManager.register)
+  def clone(session: SparkSession): ExecutionListenerManager = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: I don't know why this method is public at the first place... I have to break it here.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: Could you add MiMa exclusion rule?

**Comment by jiangxb1987** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala [hunk 3]
```diff
@@ -75,95 +76,69 @@ trait QueryExecutionListener {
  */
 @Experimental
 @InterfaceStability.Evolving
-class ExecutionListenerManager private extends Logging {
+class ExecutionListenerManager private[sql](session: SparkSession, loadExtensions: Boolean)
```
Comment (jiangxb1987) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: nit: we shall add param comments.

**Review by jiangxb1987 — APPROVED** [review]
looks good

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala (hunk 1 above)
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: Why do we want to be backwards compatible here? SHS?

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala [hunk 4]
```diff
@@ -75,95 +76,74 @@ trait QueryExecutionListener {
  */
 @Experimental
 @InterfaceStability.Evolving
-class ExecutionListenerManager private extends Logging {
-
-  private[sql] def this(conf: SparkConf) = {
-    this()
+// The `session` is used to indicate which session carries this listener manager, and we only
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: Why is this not a class doc?

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala [hunk 5]
```diff
@@ -75,95 +76,74 @@ trait QueryExecutionListener {
...
-
-  /** Acquires a write lock on the cache for the duration of `f`. */
-  private def writeLock[A](f: => A): A = {
-    val wl = lock.writeLock()
-    wl.lock()
-    try f finally {
-      wl.unlock()
-    }
+  private def shouldCatchEvent(e: SparkListenerSQLExecutionEnd): Boolean = {
+    // Only catch SQL execution with a name, and triggered by the same spark session that this
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: So this is what bugs me. You are adding separation between the SparkSession and its listeners, to undo that here. It seems like a bit of a hassle to go through because you basically need async execution.

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala [hunk 6]
```diff
@@ -75,95 +76,74 @@ trait QueryExecutionListener {
...
-  private[sql] def onSuccess(funcName: String, qe: QueryExecution, duration: Long): Unit = {
-    readLock {
-      withErrorHandling { listener =>
-        listener.onSuccess(funcName, qe, duration)
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case e: SparkListenerSQLExecutionEnd if shouldCatchEvent(e) =>
+      val funcName = e.executionName.get
+      e.executionFailure match {
+        case Some(ex) =>
+          listeners.iterator().asScala.foreach(_.onFailure(funcName, e.qe, ex))
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: This is a bit of high level thought, you could consider making the calling event queue responsible for the dispatch of these events. That way you can leverage any improvement to the underlying event bus.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: It's a developer api, which is public. The backward compatibility is not that strong, compared to end-user public APIs, but we should still keep them unchanged if not too hard.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: that said, a developer can write a spark listener and catch this event.

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: The constructor is private, so we should not make it visible in the class doc

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: `ExecutionListenerManager` is already a listener, which is running in a separated thread, receiving events from `LiveListenerBus`

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: yea. Assuming we have many spark sessions, running queries at the same time. Each session sends query execution events to the central event bus, and sets up a listener to watch its own query execution events, asynchronously.

To make it work, the most straightforward way is to carry the session identifier in the events, and the listener only watch events with the expected session identifier.

Maybe a better way is to introduce session in the Spark core, so the listener framework can dispatch events w.r.t. session automatically. But that's a lot of work.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 7]
```diff
@@ -3356,21 +3356,11 @@ class Dataset[T] private[sql](
...
-        action(qe.executedPlan)
-      }
-      val end = System.nanoTime()
-      sparkSession.listenerManager.onSuccess(name, qe, end - start)
-      result
-    } catch {
-      case e: Exception =>
-        sparkSession.listenerManager.onFailure(name, qe, e)
-        throw e
+    qe.executedPlan.foreach { plan =>
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: can this throw an exception? Imagine if `df.count()` threw an exception, and then you run it again.
Won't this be a behavior change in that case?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala [hunk 8]
```diff
@@ -71,14 +72,35 @@ object SQLExecution {
...
-          sc.listenerBus.post(SparkListenerSQLExecutionEnd(
-            executionId, System.currentTimeMillis()))
+          val endTime = System.currentTimeMillis()
+          val event = SparkListenerSQLExecutionEnd(executionId, endTime)
+          // Currently only `Dataset.withAction` and `DataFrameWriter.runCommand` specify the `name`
+          // parameter. The `ExecutionListenerManager` only watches SQL executions with name. We
+          // can specify the execution name in more places in the future, so that
+          // `QueryExecutionListener` can track more cases.
+          event.executionName = name
+          event.duration = endTime - startTime
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: duration used to be reported in nanos. Now it's millis. I would still report it as nanos if possible.

**Review by brkyvz — CHANGES_REQUESTED** [review]
This is a much larger change than I was expecting, but definitely a better one than which I had imagined. Left minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 5 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: we had the same problem in the StreamingQueryListener. You can check how we solved it in `StreamExecution`. Since each SparkSession will have its own ExecutionListenerManager, you may be able to only have the proper ExecutionListenerManager deal with its own messages.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I don't think `resetMetrics` can throw exception...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala (hunk 8 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: ah good catch!

File: sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/util/QueryExecutionListener.scala]: @brkyvz thanks for the information! It seems the `StreamingQueryListener` framework picks the same idea but the implementation is better. I'll update my PR accordingly.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 7 above)
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: can't executedPlan throw an exception? I thought it can if the original spark plan failed?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: ah i see your point here

**Comment by cloud-fan** [comment]
retest this please

**Comment by viirya** [comment]
hmm, seems it failed at the same test.

**Comment by cloud-fan** [comment]
I couldn't reproduce it locally, let me try again

**Comment by cloud-fan** [comment]
retest this please

**Comment by brkyvz** [comment]
I would just up the timeout in that suite. Now that we're pushing a bunch more stuff to the LiveListenerBus, it may not be draining quickly enough. On slow jenkins' it could likely cause flakiness.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala [hunk 9]
```diff
@@ -39,7 +39,22 @@ case class SparkListenerSQLExecutionStart(
...
-  extends SparkListenerEvent
+  extends SparkListenerEvent {
+
+  // The name of the execution, e.g. `df.collect` will trigger a SQL execution with name "collect".
+  @JsonIgnore private[sql] var executionName: Option[String] = None
+
+  // The following 3 fields are only accessed when `executionName` is defined.
+
+  // The duration of the SQL execution, in nanoseconds.
+  @JsonIgnore private[sql] var duration: Long = 0L
```
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: did you verify that the JsonIgnore annotation actually works? For some reason, I actually needed to annotate the class as
```scala
@JsonIgnoreProperties(Array("a", b", "c"))
class SomeClass {
  @JsonProperty("a") val a: ...
  @JsonProperty("b") val a: ...
}
```

the reason being Json4s understands that API better. I believe we use Json4s for all of these events

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLListener.scala]: There is a test to verify it: https://github.com/apache/spark/pull/22674/files#diff-6fa1d00d1cb20554dda238f2a3bc3ecbR55

I also used `@JsonIgnoreProperties` before, when these fields are case class fields. It seems we don't need `@JsonIgnoreProperties` when they are private `var`s.

**Comment by jiangxb1987** [comment]
LGTM, do you have any other concerns @hvanhovell @brkyvz @dongjoon-hyun ?

**Comment by cloud-fan** [comment]
since there is no objection, I'm merging it to master, thanks!
